Allow for confined users acces to wtmp

Koncpa opened the pull request:
Commit 1:
Allow for confined users acces to wtmp and run utempter
   
Allow for confined users screen acess to wtmp, via
domain transition auth_domtrans_utempter().
Utempter is specialized for updating wtmp files.

Migrated PR from selinux-policy-contrib repo.
    
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1767745

Koncpa committed:
Allow for confined users acces to wtmp and run utempter

Allow for confined users screen acess to wtmp, via
domain transition auth_domtrans_utempter().
Utempter is specialized for updating wtmp files.

Migrated PR from selinux-policy-contrib repo.

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1767745

ghost commented:
Koncpa <notifications@github.com> writes:

> Allow for confined users screen acess to wtmp, via interface
> application_exec() and auth_rw_login_records().
> Macro application_exec() allow execute application
> executables in the caller domain.
> Interface auth_rw_login_records() allow read
> and write login records.

Why not just allow screen to run utempter with a domain transition?
utempter has a private domain and utempter is specialized in using the
file.

>
> Migrated PR from selinux-policy-contrib repo.
>
> Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1767745
>
> ------------------------------------------------------------------------------------------------------
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/fedora-selinux/selinux-policy/pull/543
>
> Commit Summary
>
> * Allow for confined users acces to wtmp
>
> File Changes
>
> * M policy/modules/contrib/screen.if (3) 
>
> Patch Links:
>
> * https://github.com/fedora-selinux/selinux-policy/pull/543.patch
> * https://github.com/fedora-selinux/selinux-policy/pull/543.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

Koncpa commented:
Yes that's the option, but It's more secure to allow only necesarry permissions. 
We don't need to use domain transition and allow screen_t to have more permission than really need.

Koncpa commented:
Yes, that was a option. But is more secure to allow only permissions, which screen_t really need.

ghost commented:
Koncpa <notifications@github.com> writes:

> Yes, that was a option. But is more secure to allow only permissions, which screen_t really need.

Exactly and that is where I think you are wrong

1. youre allowing screen to execute all applications whilst it only needs
to execute utempter
2. screen *does not* need to update wtmp, utempter needs that (screen
needs to execute utempter)
3. utempters sole job is to update wtmp (it is all it does) so by allowing screen to run
utempter with a domain transition you do not allow excessive permissions 

So even though you are correct when you say "is more secure to allow
only permissions which screen_t really need". Youre making the
wrong decision.


>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

ghost commented:
I have looked into the screen policy by the way and there are some other issues that make this all less obvious.

For example currently screen is allowed to setuid and setgid unconditionally. this is less than optimal. screen can optionally be used to share sessions but this functionality is not supported by default and it requires that one manually sets the setuid bit on /usr/bin/screen.

But since this is unconditionally allowed and not a boolean it is less than obvious that if you allow screen to execute utempter without a transition then screen *needs* access to setuid/setgid *unconditionally* because utempter is setuid.

I would probably make the screen setuid/setgid access conditional with a boolean and then ensure that screen runs utempter with a domain transition

Koncpa commented:
Fine, I will look at this issue.

ghost commented:
Thanks. I just want to share knowledge. If you want help, let me know.

Koncpa commented:
So I have looked to bugzilla and It make sense your solution about utempter. So screen will be run utempter with a domain transition. But I have question about setuid and setgid.

> I have looked into the screen policy by the way and there are some other issues that make this all less obvious.
> 
> For example currently screen is allowed to setuid and setgid unconditionally. this is less than optimal. screen can optionally be used to share sessions but this functionality is not supported by default and it requires that one manually sets the setuid bit on /usr/bin/screen.
> 
Can you explain more, why setuid and setgid haven't to be allow by default? Or send me link to source, where's explanation. I need to know reason, why it should be allow only conditionally with a boolean.  Thank you

> But since this is unconditionally allowed and not a boolean it is less than obvious that if you allow screen to execute utempter without a transition then screen _needs_ access to setuid/setgid _unconditionally_ because utempter is setuid.
> 
> I would probably make the screen setuid/setgid access conditional with a boolean and then ensure that screen runs utempter with a domain transition

ghost commented:
Koncpa <notifications@github.com> writes:

> So I have looked to bugzilla and It make sense your solution about utempter. So screen will be run
> utempter with a domain transition. But I have question about setuid and setgid.
>
>  I have looked into the screen policy by the way and there are some other issues that make this all
>  less obvious.
>
>  For example currently screen is allowed to setuid and setgid unconditionally. this is less than
>  optimal. screen can optionally be used to share sessions but this functionality is not supported
>  by default and it requires that one manually sets the setuid bit on /usr/bin/screen.
>
> Can you explain more, why setuid and setgid haven't to be allow by default? Or send me link to source,
> where's explanation. I need to know reason, why it should be allow only conditionally with a boolean.
> Thank you

the setuid and setgid capbility access is only needed is the setuid bit
is set on /usr/bin/screen

do a `ls -alZ /usr/bin/screen`. Youll notice that setuid bit is not
set. This is why setuid and setgid is not needed by default. Only when
someone decides to set the setuid bit on /usr/bin/screen access to
setuid/setgid is needed, and that is why it should be considitionally
allowed with a boolean

setuid is (indirectly) needed for unix_chkpwd
(https://github.com/fedora-selinux/selinux-policy/blob/rawhide/policy/modules/contrib/screen.if#L84)
and for utempter.

but chkpwd is already run with a domain transition so that means that if
you run utempter with a domain transition, that screen_t no longer needs
CAP_SETUID/CAP_SETGID *unless* people want to share screen sessions
(which requires that the user manually sets the setuid/getgid bit on
/usr/bin/screen at which point the user should set a boolean
"screen_use_setuid_setgid" to also allow that functionality for selinux

There is *a lot* of existing bad policy in this module that can be
removed and/or fixed to improve security significantly.

does this explain it better?

>
>  But since this is unconditionally allowed and not a boolean it is less than obvious that if you
>  allow screen to execute utempter without a transition then screen needs access to setuid/setgid
>  unconditionally because utempter is setuid.
>
>  I would probably make the screen setuid/setgid access conditional with a boolean and then ensure
>  that screen runs utempter with a domain transition
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

ghost commented:
In the commit message you could put something like this:

```
screen: make setuid setgid conditional

screen session sharing requires setuid/setgid but this functionality is optional and disabled by default (the setuid/setgid bits are by default not set on /usr/bin/screen). By making the setuid and setgid capability access conditional we run screen without capability access by default and thereby improve security.

If the user wants to enable screen session sharing then the user is expected to enable this functionality by setting the setuid/setgid bits on /usr/bin/screen and to toggle the boolean "screen_allow_session_sharing" to true to allow setuid/setgid access
```

ghost commented:
https://superuser.com/a/340779

Koncpa committed:
Make for screen fsetid/setuid/setgid permission conditional

In previous screen policy was for screen sharing session
allowed fsetid/setuid/setgid capability by default. But screen session sharing is
optional funcionality, so the setuid/setgid bits
weren't set by default on screen binary. Created
screen_allow_session_sharing() boolean which, allow
use fsetid/setuid/setgid only by enabling it. These changes improves
security of screen policy.
The users can use screen session sharing by enabling this boolean
and setting manually the setuid/setgid bits on /usr/bin/screen.

Koncpa commented:
Thank you for help, so I have made commit here because is partly related with this issue.

Koncpa commented:
Commit 2: 
Make for screen fsetid/setuid/setgid permission conditional
    
In previous screen policy was for screen sharing session
allowed fsetid/setuid/setgid capability by default. But screen session sharing is
optional funcionality, so the setuid/setgid bits
weren't set by default on screen binary. Created
screen_allow_session_sharing() boolean which, allow
use fsetid/setuid/setgid only by enabling it. These changes improves
security of screen policy.
The users can use screen session sharing by enabling this boolean
and setting manually the setuid/setgid bits on /usr/bin/screen.

doverride commented on policy/modules/contrib/screen.te:
```diff
@@ -96,3 +97,7 @@ userdom_use_user_terminals(screen_domain)
 userdom_setattr_user_ttys(screen_domain)
+
+tunable_policy(`screen_allow_session_sharing',`
+allow screen_domain self:capability { setgid setuid };
```
Yes that is the gist, you could also add the fsetid here (i think the sys_tty_config can probably safely be dontaudited)

Koncpa replied:
So fsetid is somehow relates with session sharing? And about the capability of sys_tty_config, we don't know impact of this change, when we will dontaudit this capability. Sorry for late reply, but I have had a lot of work to do.

doverride replied:
Yes, capabilties require root, and root is gained with setuid. The fsetid is used to set proper ownership on the sock files used to connect to sessions

Koncpa replied:
Thank you for explanation.

ghost commented:
Koncpa <notifications@github.com> writes:

> @Koncpa commented on this pull request.
>
> ------------------------------------------------------------------------------------------------------
>
> In policy/modules/contrib/screen.te:
>
>> @@ -96,3 +97,7 @@ userdom_use_user_terminals(screen_domain)
>  userdom_create_user_pty(screen_domain)
>  userdom_setattr_user_ptys(screen_domain)
>  userdom_setattr_user_ttys(screen_domain)
> +
> +tunable_policy(`screen_allow_session_sharing',`
> +allow screen_domain self:capability { setgid setuid };
>
> So fsetid is somehow relates with session sharing? And about the capability of sys_tty_config, we
> don't know impact of this change, when we will dontaudit this capability. Sorry for late reply, but I
> have had a lot of work to do.

Yes unpriv users do not have access to capabilities (unless its ambient
i suppose). the fsetid is to
set the ownership of the sock file used to share session.

feel free to allow the sys_tty_config if you are unsure.

>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
gpg --locate-keys dominick.grift@defensec.nl
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

Koncpa commented:
> Yes unpriv users do not have access to capabilities (unless its ambient i suppose). the fsetid is to set the ownership of the sock file used to share session. feel free to allow the sys_tty_config if you are unsure.

So I finally add fsetid also to boolean and permission sys_tty_config didn't change, but build-rpm failed, so I must investigate what's happened in build.

zpytela commented:
LGTM, too, thank you.
